pkg/reconciler/cache/publishedresources/publishedresources_reconcile.go

```diff
@@ -0,0 +1,269 @@
+		ns := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        c.secretNamespace,
+				Annotations: map[string]string{logicalcluster.AnnotationKey: clusterName.String()},
+			},
+		}
+		logger = logging.WithObject(logger, ns)
+		if err := c.createNamespace(ctx, clusterName.Path(), ns); err != nil && !errors.IsAlreadyExists(err) {
+			logger.Error(err, "error creating namespace for PublishedResource secret identities")
+			// Keep going - maybe things will work. If the secret creation fails, we'll make sure to set a condition.
+		}
+	}
+}
+
+func (c *Controller) createIdentitySecret(ctx context.Context, clusterName logicalcluster.Path, apiExportName string) error {
+	secret, err := GenerateIdentitySecret(ctx, c.secretNamespace, apiExportName)
+	if err != nil {
+		return err
+	}
+	secret.Annotations[logicalcluster.AnnotationKey] = clusterName.String()
+
+	logger := logging.WithObject(klog.FromContext(ctx), secret)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(2).Info("creating identity secret")
+	return c.createSecret(ctx, clusterName, secret)
+}
+
+func (c *Controller) updateOrVerifyIdentitySecretHash(ctx context.Context, clusterName logicalcluster.Name, publishedResource *cachev1alpha1.PublishedResource) error {
+	secret, err := c.getSecret(ctx, clusterName, c.secretNamespace, publishedResource.Name)
+	if err != nil {
+		return err
+	}
+
+	hash, err := IdentityHash(secret)
+	if err != nil {
+		return err
+	}
+
+	if publishedResource.Status.IdentityHash == "" {
+		publishedResource.Status.IdentityHash = hash
+	}
+
+	if publishedResource.Status.IdentityHash != hash {
+		return fmt.Errorf("hash mismatch: identity secret hash %q must match status.identityHash %q", hash, publishedResource.Status.IdentityHash)
+	}
+
+	conditions.MarkTrue(publishedResource, cachev1alpha1.PublishedResourceIdentityValid)
+
+	return nil
+}
+
+// TODO: This is copy from apiexport controller. We should move it to a shared location.
+func IdentityHash(secret *corev1.Secret) (string, error) {
+	key := secret.Data[apisv1alpha1.SecretKeyAPIExportIdentity]
+	if len(key) == 0 {
+		return "", fmt.Errorf("secret is missing data.%s", apisv1alpha1.SecretKeyAPIExportIdentity)
+	}
+
+	hashBytes := sha256.Sum256(key)
+	hash := fmt.Sprintf("%x", hashBytes)
+	return hash, nil
+}
+
+// TODO: This is copy from apiexport controller. We should move it to a shared location.
+func GenerateIdentitySecret(ctx context.Context, ns string, name string) (*corev1.Secret, error) {
+	logger := klog.FromContext(ctx)
+	start := time.Now()
+	key := crypto.Random256BitsString()
+	if dur := time.Since(start); dur > time.Millisecond*100 {
+		logger.Info("identity key generation took a long time", "duration", dur)
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:   ns,
+			Name:        name,
+			Annotations: map[string]string{},
+		},
+		StringData: map[string]string{
+			apisv1alpha1.SecretKeyAPIExportIdentity: key,
+		},
+	}
+
+	return secret, nil
+}
```

Comment:
There already is a `pkg/crypto`, maybe this could be consolidated there?

Reply:
I think I would do small follow-up pr for this. this is gigantic already :/

Reply:
True, sounds good :)